scanpy/tools/_ingest.py

```diff
@@ -0,0 +1,239 @@
+        adata = self._adata_new
+        if self._n_pcs is not None:
+            return self._pca(self._n_pcs)
+        if self._use_rep == 'X':
+            return adata.X
+        if self._use_rep in adata.obsm.keys():
+            return adata.obsm[self._use_rep]
+        return adata.X
+
+    def transform(self, adata_new):
+        self._obs = pd.DataFrame(index=adata_new.obs.index)
+        #not sure if it should be AxisArrays
+        self._obsm = AxisArrays(adata_new, 0)
+
+        self._adata_new = adata_new
+        self._obsm['rep'] = self._same_rep()
+
+    def neighbors(self, k=10, queue_size=5, random_state=0):
+        from umap.nndescent import initialise_search
+        from umap.utils import deheap_sort
+        from umap.umap_ import INT32_MAX, INT32_MIN
+
+        random_state = check_random_state(random_state)
+        rng_state = random_state.randint(INT32_MIN, INT32_MAX, 3).astype(np.int64)
+
+        train = self._rep
+        test = self._obsm['rep']
+
+        init = initialise_search(self._rp_forest, train, test, int(k * queue_size),
+                                 self._random_init, self._tree_init, rng_state)
+
+        result = self._search(train, self._search_graph.indptr, self._search_graph.indices, init, test)
+        indices, dists = deheap_sort(result)
+        self._indices, self._distances = indices[:, :k], dists[:, :k]
+
+    def _umap_transform(self):
+        return self._umap.transform(self._obsm['rep'])
+
+    def map_embedding(self, method):
+        if method == 'umap':
+            self._obsm['X_umap'] = self._umap_transform()
+        elif method == 'pca':
+            self._obsm['X_pca'] = self._pca()
+        else:
+            raise NotImplementedError('Ingest supports only umap embeddings for now.')
+
+    def _knn_classify(self, labels):
+        cat_array = self._adata_ref.obs[labels]
+
+        values = [cat_array[inds].mode()[0] for inds in self._indices]
+        return pd.Categorical(values=values, categories=cat_array.cat.categories)
+
+    def map_labels(self, labels, method):
+        if method == 'knn':
+            self._obs[labels] = self._knn_classify(labels)
+        else:
+            raise NotImplementedError('Ingest supports knn labeling for now.')
+
+    def to_adata(self, inplace=False):
+        adata = self._adata_new if inplace else self._adata_new.copy()
+
+        adata.obsm.update(self._obsm)
+
+        for key in self._obs:
+            adata.obs[key] = self._obs[key]
+
+        if not inplace:
+            return adata
+
+    def to_adata_joint(self):
+        adata = self._adata_ref.concatenate(self._adata_new)
+
+        obs_update = self._obs.copy()
+        obs_update.index = adata[adata.obs['batch']=='1'].obs_names
+        adata.obs.update(obs_update)
+
+        for key in self._obsm:
+            if key in self._adata_ref.obsm:
+                adata.obsm[key] = np.vstack((self._adata_ref.obsm[key], self._obsm[key]))
+
+        return adata
```

Comment:
Could you document these functions (`to_adata` and `to_adata_joint`)? I think I know what they're supposed to do, but I'm not totally sure.

Reply:
Yes, i will add all needed descriptions soon.